packages/destination-actions/src/destinations/klaviyo/functions.ts

```diff
@@ -451,3 +457,177 @@ export function processPhoneNumber(initialPhoneNumber?: string, country_code?: s
+      })
+    } catch (err: any) {
+      await handleKlaviyoAPIErrorResponse(
+        transformPayloadsType(payloads),
+        await err?.response?.json(),
+        multiStatusResponse,
+        validPayloadIndicesBitmap
+      )
+    }
+  }
+
+  return multiStatusResponse
+}
+
+function validateAndPreparePayloads(payloads: TrackEventPayload[], multiStatusResponse: MultiStatusResponse) {
+  const filteredPayloads: JSONLikeObject[] = []
+  const validPayloadIndicesBitmap: number[] = []
+
+  payloads.forEach((payload, originalBatchIndex) => {
+    const { country_code, phone_number: initialPhoneNumber } = payload.profile
+
+    if (initialPhoneNumber) {
+      // Validate and convert the phone number if present
+      const validPhoneNumber = validateAndConvertPhoneNumber(initialPhoneNumber, country_code as string)
+      // If the phone number is not valid, skip this payload
+      if (!validPhoneNumber) {
+        multiStatusResponse.setErrorResponseAtIndex(originalBatchIndex, {
+          status: 400,
+          errortype: 'PAYLOAD_VALIDATION_FAILED',
+          errormessage: 'Phone number could not be converted to E.164 format.'
+        })
+        return // Skip this payload
+      }
+
+      // Update the payload's phone number with the validated format
+      payload.profile.phone_number = validPhoneNumber
+      delete payload?.profile?.country_code
+    }
+
+    // Filter out and record if payload is invalid
+    const validationError: ActionDestinationErrorResponseType | null = validatePayload(payload)
+    if (validationError) {
+      multiStatusResponse.setErrorResponseAtIndex(originalBatchIndex, validationError)
+      return
+    }
+
+    const profileToAdd = constructProfilePayload(payload)
+    filteredPayloads.push(profileToAdd as JSONLikeObject)
+    validPayloadIndicesBitmap.push(originalBatchIndex)
+    multiStatusResponse.setSuccessResponseAtIndex(originalBatchIndex, {
+      status: 200,
+      sent: profileToAdd as JSONLikeObject,
+      body: 'success'
+    })
+  })
+
+  return { filteredPayloads, validPayloadIndicesBitmap }
+}
+
+function constructProfilePayload(payload: TrackEventPayload) {
+  return {
+    type: 'event-bulk-create',
+    attributes: {
+      profile: {
+        data: {
+          type: 'profile',
+          attributes: payload.profile
+        }
+      },
+      events: {
+        data: [
+          {
+            type: 'event',
+            attributes: {
+              metric: {
+                data: {
+                  type: 'metric',
+                  attributes: {
+                    name: payload.metric_name
+                  }
+                }
+              },
+              properties: { ...payload.properties },
+              time: payload?.time ? dayjs(payload.time).toISOString() : undefined,
+              value: payload.value,
+              unique_id: payload.unique_id
+            }
+          }
+        ]
+      }
+    }
+  }
+}
+
+async function handleKlaviyoAPIErrorResponse(
```

Comment:
```suggestion
function handleKlaviyoAPIErrorResponse(
```

This function is not handling any Promises or using await, we can remove the async keyword